File: ecal/energy.js
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2025 Edvin svenblad

//============================
// Core logic for the app
//============================

//============================
//=Energy sources=============
//============================
const EType = {
  ELECTRIC:    0,
  FJARRVARME:  1,
  FJARRKYLA:   2,
  BIOBRANSLE:  3,
  FOSSIL_OLJA: 4,
  FOSSIL_GAS:  5,
  E_TYPE_COUNT: 6
};

const E_wght = [
  /* ELECTRIC   */ 1.8,
  /* FJARRVARME */ 0.7,
  /* FJARRKYLA  */ 0.6,
  /* BIOBRANSLE */ 0.6,
  /* FOSSIL_OLJA*/ 1.8,
  /* FOSSIL_GAS */ 1.8
];

const E_name = [
  /* ELECTRIC    */ "El",
  /* FJARRVARME  */ "Fjärrvärme",
  /* FJARRKYLA   */ "Fjärrkyla",
  /* BIOBRANSLE  */ "Biobränslen",
  /* FOSSIL_OLJA */ "Fossil olja",
  /* FOSSIL_GAS  */ "Fossil gas"
];

//============================
//=Common numeric constants===
//============================
const FLOW_THRESHOLD      = 0.35; // l/s·m² base limit for flow calculations
const MAX_CREDITED_FLOW   = 0.6;  // highest q_medel allowed in ep4
const EP_FLOW_FACTOR      = 40;   // multiplier for ventilation additions to EP
const EL_FLOW_BASE        = 0.022; // base coefficient for EL increase due to flow
const EL_F_GEO_MULT       = 0.02; // factor applied to F_geo in EL formulas
const EL_AREA_BASE        = 0.025; // base coefficient for EL increase due to area
const ATEMP_LIMIT         = 130;  // area limit for certain EL calculations
const AREA_THRESHOLD      = 50;   // common area threshold for multi dwelling rules
const INFILTRATION_LIMIT  = 0.60; // air leakage limit for houses without EP req

//============================
//=Limit constants============
//============================
const SMALL_LIMIT_HIGH_ATEMP = 90;  // threshold for second EP band in small houses
const SMALL_EP_OVER_ATEMP_LIMIT = 90.0; // EP limit when Atemp > ATEMP_LIMIT
const SMALL_EP_MID_ATEMP = 95.0;   // EP limit when Atemp > SMALL_LIMIT_HIGH_ATEMP
const SMALL_EP_BASE = 100.0;       // EP limit when Atemp > AREA_THRESHOLD
const SMALL_U_VALUE = 0.30;        // U-value limit for small houses
const NO_REQ_U_VALUE = 0.33;       // U-value limit when no EP requirement applies
const MULTI_EP_BASE = 75.0;        // base EP requirement for multi-dwelling houses
const MULTI_U_VALUE = 0.40;        // U-value requirement for multi-dwelling houses
const LOCAL_EP_BASE = 70.0;        // base EP requirement for local premises
const LOCAL_U_VALUE = 0.50;        // U-value requirement for local premises

//============================
//=Tappvarmvatten=============
//============================
class TvvEntry {
  constructor(name, factor, etype) {
    this.name = name;
    this.factor = factor;
    this.etype = etype;
  }
}

const tvvFactors = [
  new TvvEntry("Fjärrvärme", 1.00, EType.FJARRVARME),
  new TvvEntry("El, direktverkande och elpanna", 1.00, EType.ELECTRIC),
  new TvvEntry("El, frånluftsvärmepump", 1.70, EType.ELECTRIC),
  new TvvEntry("El, uteluft-vattenvärmepump", 2.00, EType.ELECTRIC),
  new TvvEntry("El, markvärmepump (berg, mark, sjö)", 2.50, EType.ELECTRIC),
  new TvvEntry("Biobränslepanna (pellets, ved, flis m.m.)", 0.75, EType.BIOBRANSLE),
  new TvvEntry("Olja", 0.85, EType.FOSSIL_OLJA),
  new TvvEntry("Gaspanna", 0.90, EType.FOSSIL_GAS)
];

const TvvMult = [
  /* SMALL */ 20,
  /* MULTI */ 25,
  /* LOCAL */ 2
];

//============================
//=Location===================
//============================
class Location {
  constructor(name, F_geo) {
    this.name = name;
    this.F_geo = F_geo;
  }
}

const locations = [
  new Location("Åland",   1.1),
  new Location("none",    1.0)
];

//============================
//=House Type=================
//============================
const HouseType = {
  SMALL: 0,
  MULTI: 1,
  LOCAL: 2,
  HOUSE_TYPE_COUNT: 3
};

const HouseType_name = [
  /* SMALL */ "SMALL",
  /* MULTI */ "MULTI",
  /* LOCAL */ "LOCAL"
];

//============================
//=Energy Use=================
//============================
class Energy {
  constructor() {
    this.heat = Array(EType.E_TYPE_COUNT).fill(0);
    this.cool = Array(EType.E_TYPE_COUNT).fill(0);
    this.watr = Array(EType.E_TYPE_COUNT).fill(0);
    this.fast = Array(EType.E_TYPE_COUNT).fill(0);
    this.heat_ren = Array(EType.E_TYPE_COUNT).fill(0);
    this.cool_ren = Array(EType.E_TYPE_COUNT).fill(0);
    this.watr_ren = Array(EType.E_TYPE_COUNT).fill(0);
  }
}

//============================
//=House======================
//============================
class House {
  constructor(type, Atemp, location, tvvSrc = tvvFactors[0]) {
    this.type   = type;      // HouseType
    this.Atemp  = Atemp;     // Heated area in m²
    this.E      = new Energy(); // Energy uses
    this.L      = location;  // Location object
    this.flow   = 0.0;       // Instantaneous airflow (q) [l/s·m²]
    this.qavg   = 0.0;       // Average airflow (q_medel) [l/s·m²]
    this.Rooms  = 0;         // rooms+kitchens
    this.HouseHoldEnergy = 0.0; // placeholder for household energy
    this.uval   = { U_roof: 0.0 }; // list of u-values
    this.energyusage = null; // pointer/function for usage data
    this.foot2  = false;     // Footnote‐2 flag (LOCAL)
    this.foot3  = false;     // Footnote‐3 flag (LOCAL)
    this.foot4  = false;     // Footnote‐4 flag (MULTI)
    this.foot5  = false;     // Footnote‐5 flag (MULTI)
    this.tvvSrc = tvvSrc;    // source of hot water

    // standardised tappvarmvatten energy use
    this.E.watr[tvvSrc.etype] = Tvv(this);
  }
}

//============================
//=Limits Struct==============
//============================
class LimitVals {
  constructor(EP = 0.0, EL = 0.0, UM = 0.0, LL = 0.0) {
    this.EP = EP; // Primärenergital
    this.EL = EL; // Installerad eleffekt
    this.UM = UM; // U-värde
    this.LL = LL; // Luftläckage
  }
}

//============================
//=Helper Functions============
//============================
// “No requirement” value (falls back to large number if not configured)
const NoReq = CONFIG?.CONSTANTS?.NOREQ_VALUE ?? 999999999.0;
const seSec = -1.0;

function elBase(F_geo) {
  return (4.5 + 1.7 * (F_geo - 1.0));
}

function el1(F_geo, atemp) {
  // (0,025 + 0,02 × (F_geo − 1)) × (Atemp − 130) om Atemp > 130 m²
  if (atemp <= ATEMP_LIMIT) { return 0; }
  return (EL_AREA_BASE + EL_F_GEO_MULT * (F_geo - 1.0)) * (atemp - ATEMP_LIMIT);
}

function ep2(qavg) {
  // 40 × (q_medel − 0,35) om q_medel > 0,35
  if (qavg <= FLOW_THRESHOLD) { return 0; }
  return EP_FLOW_FACTOR * (qavg - FLOW_THRESHOLD);
}

function el3(F_geo, flow, atemp) {
  // (0,022 + 0,02 × (F_geo − 1)) × (flow − 0,35) × Atemp om flow > 0,35 l/s·m²
  if (flow <= FLOW_THRESHOLD) { return 0; }
  return (EL_FLOW_BASE + EL_F_GEO_MULT * (F_geo - 1.0)) * (flow - FLOW_THRESHOLD) * atemp;
}

function ep4(F_geo, flow, qavg, atemp, Foot4) {
  // ... i flerbostadshus där Atemp är 50 m² eller större
  if (atemp < AREA_THRESHOLD) { return 0; }
  // q_medel är uteluftsflödet i temperaturreglerade utrymmen överstiger 0,35 l/s·m²
  if (qavg <= FLOW_THRESHOLD) { return 0; }
  // Tillägget kan enbart användas på grund av krav på ventilation i särskilda utrymmen som badrum, toalett och kök
  if (!Foot4) { return 0; }
  // q_medel får högst tillgodoräknas upp till 0,6 l/s·m²
  if (qavg > MAX_CREDITED_FLOW) { qavg = MAX_CREDITED_FLOW; }
  return EP_FLOW_FACTOR * (qavg - FLOW_THRESHOLD);
}

function el5(F_geo, flow, atemp, Foot5) {
  // ... i flerbostadshus där Atemp är 50 m² eller större
  if (atemp < AREA_THRESHOLD) { return 0; }
  // Tillägget kan enbart användas då det maximala uteluftsflödet vid DVUT ... överstiger 0,35 l/s·m²
  if (flow <= FLOW_THRESHOLD) { return 0; }
  // på grund av krav på ventilation i särskilda utrymmen som badrum, toalett och kök
  // och som till övervägande delen (>50 % Atemp) inrymmer lägenheter med en boarea om högst 35 m² vardera
  if (!Foot5) { return 0; }
  // Tillägg får göras med (0,022 + 0,02 × (F_geo − 1)) × (flow − 0,35) × Atemp
  return (EL_FLOW_BASE + EL_F_GEO_MULT * (F_geo - 1.0)) * (flow - FLOW_THRESHOLD) * atemp;
}

function Tvv(house) {
  return TvvMult[house.type] * house.Atemp / house.tvvSrc.factor;
}

//============================
//=EPpet Calculation===========
//============================
function EPpet(house) {
  let total = 0.0;
  const F_geo = house.L.F_geo;
  const Atemp = house.Atemp;

  for (let i = 0; i < EType.E_TYPE_COUNT; i++) {
    const heat = house.E.heat[i] - house.E.heat_ren[i];
    const cool = house.E.cool[i] - house.E.cool_ren[i];
    const watr = house.E.watr[i] - house.E.watr_ren[i];
    const fast = house.E.fast[i];
    total += ((heat / F_geo) + cool + watr + fast) * (E_wght[i] / Atemp);
  }
  return Math.floor(total);
}

//============================
//=Limits Calculation=========
//============================
function limit(house) {
  const l = new LimitVals();
  const atemp = house.Atemp;
  const F_geo = house.L.F_geo;
  let flow  = house.flow;
  const qavg = house.qavg;

  if (house.type === HouseType.SMALL) {
    if (atemp > ATEMP_LIMIT) {
      l.EP = SMALL_EP_OVER_ATEMP_LIMIT;
      l.EL = elBase(F_geo) + el1(F_geo, atemp);
      l.UM = SMALL_U_VALUE;
      l.LL = seSec;
    } else if (atemp > SMALL_LIMIT_HIGH_ATEMP) {
      l.EP = SMALL_EP_MID_ATEMP;
      l.EL = elBase(F_geo) + el1(F_geo, atemp);
      l.UM = SMALL_U_VALUE;
      l.LL = seSec;
    } else if (atemp > AREA_THRESHOLD) {
      l.EP = SMALL_EP_BASE;
      l.EL = elBase(F_geo) + el1(F_geo, atemp);
      l.UM = SMALL_U_VALUE;
      l.LL = seSec;
    } else /* atemp >= 0 */ {
      l.EP = NoReq;
      l.EL = NoReq;
      l.UM = NO_REQ_U_VALUE;
      l.LL = INFILTRATION_LIMIT;
    }
  }
  else if (house.type === HouseType.MULTI) {
    if (atemp >= 0) {
      l.EP = MULTI_EP_BASE + ep4(F_geo, flow, qavg, atemp, house.foot4);
      l.EL = elBase(F_geo)
             + el1(F_geo, atemp)
             + el5(F_geo, flow, atemp, house.foot5);
      l.UM = MULTI_U_VALUE;
      l.LL = seSec;
    }
  }
  else if (house.type === HouseType.LOCAL) {
    if (atemp > AREA_THRESHOLD) {
      // Only add ep2(qavg) if foot2 is checked
      const ep_add = house.foot2 ? ep2(qavg) : 0;
      // Only add el3(...) if foot3 is checked
      const el_add = house.foot3 ? el3(F_geo, flow, atemp) : 0;

      l.EP = LOCAL_EP_BASE + ep_add;
      l.EL = elBase(F_geo)
             + el1(F_geo, atemp)
             + el_add;
      l.UM = LOCAL_U_VALUE;
      l.LL = seSec;
    } else /* atemp >= 0 */ {
      l.EP = NoReq;
      l.EL = NoReq;
      l.UM = NO_REQ_U_VALUE;
      l.LL = INFILTRATION_LIMIT;
    }
  }

  return l;
}


// Export for Node.js testing
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    EType,
    E_wght,
    E_name,
    Location,
    locations,
    HouseType,
    HouseType_name,
    Energy,
    House,
    LimitVals,
    elBase,
    el1,
    ep2,
    el3,
    ep4,
    el5,
    EPpet,
    limit
  };
}

File: ecal/dev/energy.c


#include <stdio.h>
#include <string.h>

#define ef        else if
#define I         int
#define D         double
#define R         return

typedef struct {
        char name[256];
        D    factor;
} EntryD;

typedef struct {
        char name[256];
        D    F_geo;
} Location;

typedef enum {
	SMALL,
	MULTI,
	LOCAL,
	HOUSE_TYPE_COUNT
} HouseType;
typedef enum {
	SMALL_, MULTI_, MULTI_OLD, LOCAL_OFFICE, LOCAL_PRESCOOL, LOCAL_SCHOOL, LOCAL_UNI,
	HOUSE__SUB_TYPE_COUNT
} HouseSubType;
const char *HouseType_name[HOUSE_TYPE_COUNT] = {
	[SMALL] = "SMALL",
	[MULTI] = "MULTI",
	[LOCAL] = "LOCAL"
};
//============================
//=Energy sources=============
//============================

typedef enum {
	ELECTRIC,
	FJARRVARME,
	FJARRKYLA,
	BIOBRANSLE,
	FOSSIL_OLJA,
	FOSSIL_GAS,
	E_TYPE_COUNT
} EType;

static const EntryD E_types[E_TYPE_COUNT] = {
        [ELECTRIC]    = { "El",          1.8 },
        [FJARRVARME]  = { "Fjärrvärme",  0.7 },
        [FJARRKYLA]   = { "Fjärrkyla",   0.6 },
        [BIOBRANSLE]  = { "Biobränslen", 0.6 },
        [FOSSIL_OLJA] = { "Fossil olja", 1.8 },
        [FOSSIL_GAS]  = { "Fossil gas",  1.8 }
};

typedef struct {
        char  name[256];
        D     factor;
        EType etype;
} TvvEntry;

typedef struct UVals {
        D     U_roof;
        ///...
} UVals;


static const Location locations[] = {
	{ "Åland",   1.1 },
	{ "none", 1.0 },
};
static const TvvEntry tvvFactors[] = {
        { .name = "Fjärrvärme", .factor = 1.00, .etype = FJARRVARME },
        { .name = "El, direktverkande och elpanna", .factor = 1.00, .etype = ELECTRIC },
        { .name = "El, frånluftsvärmepump", .factor = 1.70, .etype = ELECTRIC },
        { .name = "El, uteluft-vattenvärmepump", .factor = 2.00, .etype = ELECTRIC },
        { .name = "El, markvärmepump (berg, mark, sjö)", .factor = 2.50, .etype = ELECTRIC },
        { .name = "Biobränslepanna (pellets, ved, flis m.m.)", .factor = 0.75, .etype = BIOBRANSLE },
        { .name = "Olja", .factor = 0.85, .etype = FOSSIL_OLJA },
        { .name = "Gaspanna", .factor = 0.90, .etype = FOSSIL_GAS }
};
const double TvvMult[HOUSE_TYPE_COUNT] = {
        [SMALL] = 20,
        [MULTI] = 25,
        [LOCAL] = 2,
};




//============================
//=Energy Use=================
//============================
typedef struct {
        D heat[E_TYPE_COUNT];
        D cool[E_TYPE_COUNT];
        D watr[E_TYPE_COUNT];
        D fast[E_TYPE_COUNT];
        D heat_ren[E_TYPE_COUNT];
        D cool_ren[E_TYPE_COUNT];
        D watr_ren[E_TYPE_COUNT];
} Energy;

//============================
//=House======================
//============================
typedef struct {
        HouseType      type;
        I              Atemp;      // Heated area in m²
        Energy         E;          // Energy uses
        const Location *L;       // location pointer
        D              flow;       // Instantaneous airflow (q) [l/s·m²]
        D              qavg;       // Average airflow (q_medel) [l/s·m²]
        TvvEntry     tvvSrc;	//source of hot water
	I 		Rooms;	//rooms+kitchens

	D 		HouseHoldEnergy; //stubb


	UVals uval;		//list of uvalues
	I* 		energyusage; //pointer to function?
        I              foot2;      // Footnote-2 flag
        I              foot3;      // Footnote-3 flag
        I              foot4;      // Footnote-4 flag
        I              foot5;      // Footnote-5 flag
} House;

const int PERSON_HEAT = 80; // watts

//returns the calculated tvv value for the selected building type
D Tvv(House H) {
        R TvvMult[H.type] * H.Atemp / H.tvvSrc.factor;
}

//============================
//=Limits Struct==============
//============================
typedef struct {
	D EP;   // Primärenergital
	D EL;   // Installerad eleffekt
	D UM;   // U-värde
	D LL;   // Luftläckage
} LimitVals;

//============================
//=Helper Functions============
//============================
#define NoReq  9999.0
#define seSec  -1.0

D elBase(D F_geo) { R (4.5 + 1.7 * (F_geo - 1.0)); }

D el1(D F_geo, I atemp) {
	// (0,025 + 0,02 × (F_geo − 1)) × (Atemp − 130) om Atemp > 130 m²
	if (atemp <= 130) { R 0; }
	R (0.025 + 0.02 * (F_geo - 1.0)) * (atemp - 130);
}

D ep2(D qavg) {
	// 40 × (q_medel − 0,35) om q_medel > 0,35
	if (qavg <= 0.35) { R 0; }
	R 40 * (qavg - 0.35);
}

D el3(D F_geo, D flow, I atemp) {
	// (0,022 + 0,02 × (F_geo − 1)) × (flow − 0,35) × Atemp om flow > 0,35 l/s·m²
	if (flow <= 0.35) { R 0; }
	R (0.022 + 0.02 * (F_geo - 1.0)) * (flow - 0.35) * atemp;
}

D ep4(D F_geo, D flow, D qavg, I atemp, I Foot4) {
	// ...i flerbostadshus där Atemp är 50 m² eller större
	if (atemp < 50) { R 0; }
	// q_medel är uteluftsflödet i temperaturreglerade utrymmen överstiger 0,35 l/s·m²
	if (qavg <= 0.35) { R 0; }
	// Tillägget kan enbart användas på grund av krav på ventilation i särskilda utrymmen som badrum, toalett och kök
	if (!Foot4) { R 0; }
        // q_medel får högst tillgodoräknas upp till 0,6 l/s·m²
        if (qavg > 0.6) { qavg = 0.6; }
	R 40 * (qavg - 0.35);
}

D el5(D F_geo, D flow, I atemp, I Foot5) {
	// ...i flerbostadshus där Atemp är 50 m² eller större
	if (atemp < 50) { R 0; }
	// Tillägget kan enbart användas då det maximala uteluftsflödet vid DVUT i temperaturreglerade utrymmen q överstiger 0,35 l/s·m²
	if (flow <= 0.35) { R 0; }
	// på grund av krav på ventilation i särskilda utrymmen som badrum, toalett och kök
	// och som till övervägande delen (>50 % Atemp) inrymmer lägenheter med en boarea om högst 35 m² vardera
	if (!Foot5) { R 0; }
	// Tillägg får göras med (0,022 + 0,02 × (F_geo − 1)) × (flow − 0,35) × Atemp
	R (0.022 + 0.02 * (F_geo - 1.0)) * (flow - 0.35) * atemp;
}

//============================
//=Function Prototypes========
//============================
House     newHouse(HouseType type, I atemp, const Location *L, const TvvEntry *tvvSrc);
I         EnergyUsage(const House *h);
I         EPpet(const House *h);
LimitVals limit(const House *h);
void      printHouse(const House *h);

//============================
//=Constructor================
//============================

House newHouse(HouseType type, I atemp, const Location *L, const TvvEntry *tvvSrc) {
	House h;
	h.type   = type;
	h.Atemp  = atemp;
	memset(&h.E, 0, sizeof(Energy));
	h.L      = L;

        h.flow   = 0.0;
        h.qavg   = 0.0;
        h.Rooms  = 0;
        h.HouseHoldEnergy = 0.0;
        memset(&h.uval, 0, sizeof(UVals));
        h.energyusage = NULL;

        h.foot4  = 0;  h.foot2  = 0;
        h.foot3  = 0;
        h.foot5  = 0;
        h.tvvSrc = *tvvSrc;
        h.E.watr[tvvSrc->etype] = Tvv(h);
        R h;

}

//============================
//=Main=======================
//============================
I main() {
	// Create a House (Åland, SMALL, Atemp = 100)
     House h = newHouse(SMALL, 100, &locations[0], &tvvFactors[0]);

	// Example inputs:
	h.E.heat[ELECTRIC]   = 120.0;
	h.E.cool[FJARRKYLA]  = 40.0;
	h.flow               = 0.5;   // example instantaneous airflow (q)
	h.qavg               = 0.4;   // example average airflow (q_medel)
	h.foot4              = 1;     // footnote-4 applies
	h.foot5              = 1;     // footnote-5 applies

	printHouse(&h);
	R 0;

}

//============================
//=Core Calculation===========
//============================
I EPpet(const House *h) {
	D total = 0.0;
	D F_geo = h->L->F_geo;
	I  Atemp = h->Atemp;

        for (I i = 0; i < E_TYPE_COUNT; i++) {
                total += (
                                ((h->E.heat[i] - h->E.heat_ren[i]) / F_geo
                                 + (h->E.cool[i] - h->E.cool_ren[i])
                                 + (h->E.watr[i] - h->E.watr_ren[i])
                                 + h->E.fast[i])
                                * (E_types[i].factor / Atemp)
                         );
        }
	R (I)total;
}

//============================
//=Limits Calculation=========
//============================
LimitVals limit(const House *h) {
	LimitVals l = {0.0, 0.0, 0.0, 0.0};
	I      atemp = h->Atemp;
	D      F_geo = h->L->F_geo;
	D      flow  = h->flow;
	D      qavg  = h->qavg;

        if (atemp <= 0) { R l; }

	if (h->type == SMALL) {
		if (atemp > 130) {
			l.EP = 90.0;
			l.EL = elBase(F_geo) + el1(F_geo, atemp);
			l.UM = 0.30;
			l.LL = seSec;
		}
		ef (atemp > 90) {
			l.EP = 95.0;
			l.EL = elBase(F_geo) + el1(F_geo, atemp);
			l.UM = 0.30;
			l.LL = seSec;
		}
		ef (atemp > 50) {
			l.EP = 100.0;
			l.EL = elBase(F_geo) + el1(F_geo, atemp);
			l.UM = 0.30;
			l.LL = seSec;
		}
		ef (atemp >= 0) {
			l.EP = NoReq;
			l.EL = NoReq;
			l.UM = 0.33;
			l.LL = 0.60;
		}
	}
	ef (h->type == MULTI) {
		if (atemp >= 0) {
			l.EP = 75.0 + ep4(F_geo, flow, qavg, atemp, h->foot4);
			l.EL = elBase(F_geo) + el1(F_geo, atemp) + el5(F_geo, flow, atemp, h->foot5);
			l.UM = 0.40;
			l.LL = seSec;
		}
	}
        ef (h->type == LOCAL) {
                if (atemp > 50) {
                        D ep_add = h->foot2 ? ep2(qavg) : 0.0;
                        D el_add = h->foot3 ? el3(F_geo, flow, atemp) : 0.0;
                        l.EP = 70.0 + ep_add;
                        l.EL = elBase(F_geo) + el1(F_geo, atemp) + el_add;
                        l.UM = 0.50;
                        l.LL = seSec;
                }
                ef (atemp >= 0) {
                        l.EP = NoReq;
                        l.EL = NoReq;
                        l.UM = 0.33;
                        l.LL = 0.60;
                }
        }
	R l;
}

//============================
//=Print House Data===========
//============================
void printHouse(const House *h) {
        printf("House type: %s\n", HouseType_name[h->type]);
        printf("Atemp: %d m²\n", h->Atemp);
        printf("location: %s (F_geo = %.2f)\n", h->L->name, h->L->F_geo);
        printf("Rooms: %d\n", h->Rooms);
        printf("Household energy: %.1f\n", h->HouseHoldEnergy);
        printf("U_roof: %.2f\n", h->uval.U_roof);
        printf("Flow (q): %.2f   qavg (q_medel): %.2f\n", h->flow, h->qavg);
        printf("Foot2: %s   Foot3: %s   Foot4: %s   Foot5: %s\n",
                        h->foot2 ? "Yes" : "No",
                        h->foot3 ? "Yes" : "No",
                        h->foot4 ? "Yes" : "No",
                        h->foot5 ? "Yes" : "No");

	printf("Energy use:\n");
	for (I i = 0; i < E_TYPE_COUNT; i++) {
		if (h->E.heat[i] != 0.0)
			printf("  %s heat: %.1f  ", E_types[i].name, h->E.heat[i]);
		if (h->E.cool[i] != 0.0)
			printf("  %s cool: %.1f  ", E_types[i].name, h->E.cool[i]);
		if (h->E.watr[i] != 0.0)
			printf("  %s watr: %.1f  ", E_types[i].name, h->E.watr[i]);
		if (h->E.fast[i] != 0.0)
			printf("  %s fast: %.1f  ", E_types[i].name, h->E.fast[i]);
		if (h->E.heat[i] != 0.0 ||
				h->E.cool[i] != 0.0 ||
				h->E.watr[i] != 0.0 ||
				h->E.fast[i] != 0.0)
			printf("\n");
	}


	I ep = EPpet(h);
	printf("Calculated EP: %d\n", ep);

	LimitVals lim = limit(h);
	printf("Limits → EP: %.1f, EL: %.1f, UM: %.2f, LL: %.2f\n",
			lim.EP, lim.EL, lim.UM, lim.LL);
}

